docs: ADR-0033 feature-usage bitmask in the User-Agent (#6500)
microsoft/agent-framework: docs/decisions/0033-feature-usage-bitmask-user-agent.md

- Neutral, two small lists to maintain instead of one (but they were going to
  diverge anyway — the packages differ).

#### I. Single shared cross-language registry

- Good, one list, one number space.
- Bad, forces synchronized numbering and null placeholders for features that
  exist in only one SDK, plus SDK-aware decode rules.
- Bad, the synchronization is pure accidental complexity — **the language is
  already in the User-Agent**, so sharing the number space buys nothing.

### Registry maintenance

#### J. Hand-written enum + parity test (chosen)

- Good, ~40 members that change a few times a year; a 10-line test (the enum vs
  the per-language table in the registry doc) is enough.
- Good, no build step, no generator to own.

#### K. Code-generate the enums from the registry

- Bad, a generator + drift test + schema test to maintain a short list of
  integer constants; justified only by the per-construct bit count we rejected.

### Representation (how the mask is rendered as text)

All examples below encode the same mask — bits 0, 2, 16, 22, 27 set
(agent + workflow + sequential-orchestration + foundry.chat_client + openai, in

> the Python v1 list) = decimal `138477573`.

Reviewer: Have we gone over how the registry is versioned?

Author: there is a section on that topic, the basics are to only up the version when we need to remove keys, adding can be done inplace (within the limits of the number of bits)

  the marker API exists.
- Good, one repository test compares the package-local declarations with the
  per-language table and rejects missing rows, wrong ids, out-of-range indexes,
  and any duplicate/overlapping index.
- Good, no build step, no generator to own.

#### K. Code-generate the enums from the registry

- Bad, a generator + drift test + schema test to maintain a short list of
  integer constants; likely justified only if v1 deliberately chooses the most
  detailed L3/L4 granularities.

### Representation (how the mask is rendered as text)

All examples below encode the same mask — bits 0, 2, 32, 48, 56 set
(agent + workflow + sequential-orchestration + foundry.chat_client + openai, in
the Python v1 list) = decimal `72339073309605893`.

#### L. Decimal — `feat=v1.72339073309605893`

- Good, human-familiar; trivial to parse.
- Neutral, no visual alignment to bit/nibble boundaries; slightly longer than hex
  for large masks. No advantage over hex.

#### M. Hex (chosen) — `feat=v1.101000100000005`

- Good, compact (≤32 chars for a 128-bit mask).
- Good, decodes with one stdlib call in every language (`int(x, 16)` /

>   two 64-bit lane parses in .NET); nibble boundaries are eyeball-able.

Reviewer 2: Suggest removing any jargons from the ADR. "nibble boundaries are eyeball-able" ???

Also, suggest another option where the "nibble" boundaries are separated by a "." separator char, for ease of looking, possibly something that could be compatible with a `int(x, 16)` call with the benefit of extra visibility.

```suggestion
  two 64-bit lane parses in .NET); and because each character always stands for the same group of four bits, you can read off which features are set just by looking.
```

Author: Addressed in 982a92cd7. The jargon was removed and the text now says directly that each hex character maps to four consecutive bit positions. A dotted grouped-hex form is documented as considered but rejected because separators add bytes and must be stripped before standard hex parsing.

- Good, lowercase, no `0x` prefix, no leading zeros — unambiguous and stable.

#### N. Binary / bit-list — `feat=v1.100000001000000000000000100000000000000000000000000000101` or `feat=v1.0,2,32,48,56`

- Good, most directly human-readable ("which bits").
- Bad, longest form in the UA; the bit-list needs delimiter handling and grows

>   with the number of set bits.

Reviewer 2: Bit list should be separated from the binary option.

I `feat=v1.0,1,2,3,4,5` is clearly eyeballing transparent while, this option should also come with strong recommendations on how many bits should be present in a single call at the same time. Do we have an AVG estimate, are you thinking in over using the bits?

What is extra the processing cost of adding the bits detection for building up the UA?

Would be nice also to provide what is a char count in a 128 bits scenario for every option so we have a table of how many bytes are expected to be added per request

Are we thinking in providing up to 128 bits of information in our calls?

Author: Addressed in 982a92cd7. Binary and bit-list are now separate alternatives. The ADR adds exact added-User-Agent byte counts for the example, all currently assigned Python/.NET rows, and the full 128-bit bound; it also states that there is no defensible average or smaller per-request bit budget and describes the bounded marking/stamping work.

  eyeball-able. Premature optimization for a value that is already ≤32 chars in
  hex.

## Decision Outcome

Chosen: **a per-request, first-party-only User-Agent `(feat=...)` token (A),
with a 128-bit process-global monotonic accumulator (S1), per-language bit lists
(H), package-local index enums kept honest by parity and no-overlap tests (J),
rendered as lowercase hex (M).**

This is a bounded design with enough v1 headroom. A 128-bit
**process-global, monotonic** mask accumulates from universal
`mark_feature_used()` calls (so it spans build/start/participation activations
that aren't bound to any service request — the per-request set model (S2) can't);
the token is **stamped per request** only when both the client/pipeline and the
actual HTTPS origin are approved, so custom origins and cross-origin redirects
cannot inherit the fingerprint; each
SDK owns an independent bit list selected by the language already in the UA; the
mask is rendered as hex (`feat=v1.101000100000005`). **Two opt-out env vars are
provided:** a dedicated `AGENT_FRAMEWORK_FEATURE_MASK_DISABLED` that drops only
the mask while keeping the base SDK identity/version User-Agent, and the existing

> `AGENT_FRAMEWORK_USER_AGENT_DISABLED` that drops the whole contribution. OTel (C)

Reviewer 2: I'm not sure we do have proposed or existing anything that drops the MAF - UA stamping, not sure if this should be added as part of this ADR, suggest limiting the env var toggling for the Feature Bit Stamping, in the scope of this PR.

```suggestion
the mask while keeping the base SDK identity/version User-Agent
```

Author: Addressed in 982a92cd7. This ADR now introduces only the dedicated mask-only flag across both SDKs. The existing Python whole-User-Agent flag still suppresses the Python mask, while adding an equivalent .NET whole-UA switch is explicitly outside this decision.

decimal/binary/base-N representations (L, N, O) are rejected as complexity or
length the problem does not require.

The remaining choice before implementation is the **v1 granularity level** among
F0-F4. This is a point-in-time decision: adding new bits later is easier than
removing or redefining them, because removals/redefinitions require a new
registry version and historical decode tables. For v1, prefer the least detailed
level that answers the known product/support questions so we do not force a v2
shortly after launch. The refreshed candidate registry uses **63 Python indexes and
52 .NET indexes**, leaving 65 and 76 positions respectively. That headroom supports
normal growth; it does not waive the registry's
[allocation tenet](../specs/feature-usage-bit-registry.md#allocation-tenet).

### Consequences

- Good, adds usage signal at near-zero cost, no new data flow, few moving parts.
- Good, transparent (public registry, human-decodable token) and disabled by
  **two** opt-out env vars: a dedicated `AGENT_FRAMEWORK_FEATURE_MASK_DISABLED`
  (mask only) and `AGENT_FRAMEWORK_USER_AGENT_DISABLED` (whole UA; existing in

>   Python and added to .NET with this work).

Reviewer 2: See above.

```suggestion
- Good, transparent (public registry, human-decodable token) and disabled by
  **one** opt-out env vars: a dedicated `AGENT_FRAMEWORK_FEATURE_MASK_DISABLED`
  (mask only).
```

Author: Addressed in 982a92cd7. The outcome and consequences now describe one new `AGENT_FRAMEWORK_FEATURE_MASK_DISABLED` opt-out. The pre-existing Python whole-UA behavior is noted separately rather than standardized as new cross-language behavior.
